### PR TITLE
nixos/borgbackup: Add a persistentTimer option.

### PR DESCRIPTION
**Pull request by Stunkymonkey**
Persistent starts the backup service on power on if it was missed while
the system was powered down, for example.

###### Motivation for this change
There have been multiple tries to get the persistent option for borgbackup (#75056, #99196, #123434). Here I am trying again to get a combination of all PRs into one, that will get accepted.
This is my first nixos-option i am providing, so read carefully.

I decided to default to `false`, because that is the current state. Not sure if enabling would break something.
But i am fine in changing this, because most of the users would change it to `true`.

Other then that there is one improvement left to be done: when using WiFi the network is not yet connected and the backup would always fail. Maybe we could improve it by checking if the `repo` contains an `@` or `:` symbol. If this is the case then we add:
```
wants = [ "network-online.target" ];
after = [ "network-online.target" ];
```
In the [wiki](https://nixos.wiki/wiki/Borg_backup) they do a ping to google, which i am not a big fan of.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

ping @dotlambda, @Ma27, @danbst, @ngiger, @Mic92

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
@dotlambda Can you please take a look at this?

**Comment by Stunkymonkey** [comment]
@dotlambda the first commit it tested with multiple configurations.
the second one I just added, because I think it makes sense to wait for the network and then start the timer.

File: nixos/modules/services/backup/borgbackup.nix [hunk 1]
```diff
@@ -102,6 +102,16 @@ let
       inherit (cfg) startAt;
```
Comment (pennae) [inline nixos/modules/services/backup/borgbackup.nix:102]: would prefer that non-`systemd.timers` timer config goes away if we also use `systemd.timers` config. it's easier to understand when all information for a thing is in one place.

File: nixos/modules/services/backup/borgbackup.nix [hunk 2]
```diff
@@ -320,6 +330,19 @@ in {
             '';
           };
 
+          persistentTimer = mkOption {
+            default = false;
+            type = types.bool;
+            example = true;
+            description = ''
```
Comment (pennae) [inline nixos/modules/services/backup/borgbackup.nix:337]:
```suggestion
            description = literalDocBook ''
```

File: nixos/modules/services/backup/borgbackup.nix (hunk 1 above)
Comment (Stunkymonkey) [reply nixos/modules/services/backup/borgbackup.nix:102]: you mean moving this line down to the timer part?

Comment (pennae) [reply nixos/modules/services/backup/borgbackup.nix:102]: yup, all it does now anyway is add an `OnCalendar` entry to the timer.

**Review by pennae — COMMENTED** [review]
looks good, let's just have the tests run and then go.

@ofborg test borgbackup

**Comment by pennae** [comment]
since doing this in a review comment apparently hasn't worked:

@ofborg test borgbackup

File: nixos/modules/services/backup/borgbackup.nix [hunk 3]
```diff
@@ -99,7 +99,17 @@ let
...
       } // (mkPassEnv cfg) // cfg.environment;
-      inherit (cfg) startAt;
+    };
+
+  mkBackupTimers = name: cfg:
+    nameValuePair "borgbackup-job-${name}" {
+      description = "BorgBackup job ${name} timer";
+      timerConfig = {
+        Persistent = cfg.persistentTimer;
+        OnCalendar = cfg.startAt;
```
Comment (pennae) [inline nixos/modules/services/backup/borgbackup.nix:109]: sorry to be at this again, but having tested a few things locally this no longer looks correct. systemd refuses timer units with no triggers (which could happen if `startAt == []`, which is totally valid). we should probably generate a timer unit only for backup jobs that have `startAt != []`. :slightly_frowning_face:

Comment (Stunkymonkey) [reply nixos/modules/services/backup/borgbackup.nix:109]: having `startAt == []` means the service is available but has to be triggered manually?

Comment (pennae) [reply nixos/modules/services/backup/borgbackup.nix:109]: either that, or you can use the `borg-job-${name}` wrapper that is provided for each configured job

Comment (Stunkymonkey) [reply nixos/modules/services/backup/borgbackup.nix:109]: sorry, but this is my first nixos-option i am doing. I might need some more guidance on how to do that.

could we use something like `(filterAttrs (cfg: cfg.startAt != []) cfg.services)` or what are more fitting functions?

should the filtering be done at line 104? or in 722?

Comment (pennae) [reply nixos/modules/services/backup/borgbackup.nix:109]: you could filter in either place, but in 722 it makes for a definition that's a bit easier to understand. in 104 you'd have to wrap the value in a `mkIf (cfg.startAt != [])`, which looks really odd and like spooky action at a distance :) in 722 it'd just be changing `jobs` to `(filterAttrs (cfg: cfg.startAt != []) jobs)`, which seems preferrable for clarity of what happens.

Comment (Stunkymonkey) [reply nixos/modules/services/backup/borgbackup.nix:109]: i added a little comment as well

Comment (pennae) [reply nixos/modules/services/backup/borgbackup.nix:109]: looks good :)

**Review by pennae — APPROVED** [review]

**Comment by Stunkymonkey** [comment]
@ofborg test borgbackup

**Comment by Stunkymonkey** [comment]
should i replace `cfg.startAt != []` with `cfg.startAt != ""`? does that fix it?

File: nixos/modules/services/backup/borgbackup.nix [hunk 4]
```diff
@@ -695,6 +718,10 @@ in {
         # A repo named "foo" is mapped to systemd.services.borgbackup-repo-foo
         // mapAttrs' mkRepoService repos;
 
+      # A job named "foo" is mapped to systemd.timers.borgbackup-job-foo
+      # only generate the timer if interval (startAt) is set
+      systemd.timers = mapAttrs' mkBackupTimers (filterAttrs (cfg: cfg.startAt != []) jobs);
```
Comment (pennae) [inline nixos/modules/services/backup/borgbackup.nix:723]:
```suggestion
      systemd.timers = mapAttrs' mkBackupTimers (_: filterAttrs (cfg: cfg.startAt != []) jobs);
```

Comment (pennae) [inline nixos/modules/services/backup/borgbackup.nix:723]:
```suggestion
      systemd.timers = mapAttrs' mkBackupTimers (filterAttrs (_: cfg: cfg.startAt != []) jobs);
```

Comment (pennae) [reply nixos/modules/services/backup/borgbackup.nix]: ignore the first suggestion, put the argument in the wrong place by accident :sweat:

**Comment by Stunkymonkey** [comment]
@ofborg test borgbackup

File: nixos/modules/services/backup/borgbackup.nix (hunk 4 above)
Comment (Stunkymonkey) [reply nixos/modules/services/backup/borgbackup.nix]: still does not work :disappointed:

Comment (pennae) [reply nixos/modules/services/backup/borgbackup.nix]: we mistyped the suggestion at first, looks like you applied that wrong one. we tested locally with this one and it worked out :)

**Comment by Stunkymonkey** [comment]
@ofborg test borgbackup

File: nixos/modules/services/backup/borgbackup.nix [hunk 5]
```diff
@@ -99,7 +99,17 @@ let
         BORG_REPO = cfg.repo;
         inherit (cfg) extraArgs extraInitArgs extraCreateArgs extraPruneArgs;
       } // (mkPassEnv cfg) // cfg.environment;
-      inherit (cfg) startAt;
+    };
+
+  mkBackupTimers = name: cfg:
+    nameValuePair "borgbackup-job-${name}" {
+      description = "BorgBackup job ${name} timer";
```
Comment (pennae) [inline nixos/modules/services/backup/borgbackup.nix:106]: timers are correct now but aren't started automatically anymore :( gosh i hate computers sometimes, moving this one line was such a bad idea on the effort vs usefulness scale, so sorry for all the trouble :sweat: 
```suggestion
      description = "BorgBackup job ${name} timer";
      wantedBy = [ "timers.target" ];
```

Comment (Stunkymonkey) [reply nixos/modules/services/backup/borgbackup.nix:106]: thanks for taking so much time. initially i just wanted to have a persistent method, but i fell this is going into a very good direction.

**Comment by pennae** [comment]
@ofborg test borgbackup

**Comment by blargg** [comment]
Hey, thanks for putting this together! I appreciate it. I must have lost the thread on my PR, and just noticed when I started to hit conflicts with my weird setup.

**Pull request merged**